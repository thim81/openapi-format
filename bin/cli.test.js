const testUtils = require('../test/__utils__/test-utils')
const fs = require("fs");

describe("openapi-format CLI command", () => {

  it("should output the help", async () => {
    let result = await testUtils.cli([`--help`], '.');
    // console.log('result', result)
    expect(result.code).toBe(0);
    expect(result.stderr).toMatchSnapshot();
  });

  it("should use the default sort", async () => {
    const path = `test/yaml-default`
    const inputFile = `${path}/input.yaml`
    const outputFile = `${path}/output.yaml`
    const output = (fs.readFileSync(outputFile, 'utf8'));

    let result = await testUtils.cli([inputFile], '.');
    // console.log('result', result)
    expect(result.code).toBe(0);
    expect(result.stdout).toContain("formatted successfully");
    expect(result.stdout).toMatchSnapshot();
    expect(sanitize(result.stderr)).toStrictEqual(sanitize(output));
  });

  it("should stop and show error", async () => {
    const path = `test/yaml-default`
    const inputFile = `${path}/input.yaml`
    const outputFile = `${path}/output.yaml`

    let result = await testUtils.cli([inputFile, `--sortFile foobar`], '.');
    // console.log('result', result)
    expect(result.code).toBe(1);
    expect(result.stdout).toMatchSnapshot();
  });

  it("should use the sortFile", async () => {
    const path = `test/yaml-custom`
    const inputFile = `${path}/input.yaml`
    const outputFile = `${path}/output.yaml`
    const output = (fs.readFileSync(outputFile, 'utf8'));
    const setting = `${path}/customSort.yaml`

    let result = await testUtils.cli([inputFile, `--sortFile ${setting}`], '.');
    // console.log('result', result)
    expect(result.code).toBe(0);
    expect(result.stdout).toContain("formatted successfully");
    expect(result.stdout).toMatchSnapshot();
    expect(sanitize(result.stderr)).toStrictEqual(sanitize(output));
  });

  it("should use the casingFile", async () => {
    const path = `test/yaml-casing`
    const inputFile = `${path}/input.yaml`
    const outputFile = `${path}/output.yaml`
    const output = (fs.readFileSync(outputFile, 'utf8'));
    const setting = `${path}/customCasing.yaml`

    let result = await testUtils.cli([inputFile, `--casingFile ${setting}`, `--no-sort`], '.');
    // console.log('result', result)
    expect(result.code).toBe(0);
    expect(result.stdout).toContain("formatted successfully");
    expect(result.stdout).toMatchSnapshot();
    expect(sanitize(result.stderr)).toStrictEqual(sanitize(output));
  });

  it("should use the filterFile", async () => {
    const path = `test/yaml-filter-custom`
    const inputFile = `${path}/input.yaml`
    const outputFile = `${path}/output.yaml`
    const output = (fs.readFileSync(outputFile, 'utf8'));
    const setting = `${path}/customFilter.yaml`

    let result = await testUtils.cli([inputFile, `--filterFile ${setting}`, `--no-sort`], '.');
    // console.log('result', result)
    expect(result.code).toBe(0);
    expect(result.stdout).toContain("formatted successfully");
    expect(result.stdout).toMatchSnapshot();
    expect(sanitize(result.stderr)).toStrictEqual(sanitize(output));
  });

  it("should use the sortComponentsFile", async () => {
    const path = `test/yaml-sort-components`
    const inputFile = `${path}/input.yaml`
    const outputFile = `${path}/output.yaml`
    const output = (fs.readFileSync(outputFile, 'utf8'));
    const setting = `${path}/customSortComponents.yaml`

    let result = await testUtils.cli([inputFile, `--sortComponentsFile ${setting}`], '.');
    // console.log('result', result)
    expect(result.code).toBe(0);
    expect(result.stdout).toContain("formatted successfully");
    expect(result.stdout).toMatchSnapshot();
    expect(sanitize(result.stderr)).toStrictEqual(sanitize(output));
  });

  it("should use the no-sort", async () => {
    const path = `test/yaml-no-sort`
    const inputFile = `${path}/input.yaml`
    const outputFile = `${path}/output.yaml`
    const output = (fs.readFileSync(outputFile, 'utf8'));

    let result = await testUtils.cli([inputFile, `--no-sort`], '.');
    // console.log('result', result)
    expect(result.code).toBe(0);
    expect(result.stdout).toContain("formatted successfully");
    expect(result.stdout).toMatchSnapshot();
    expect(sanitize(result.stderr)).toStrictEqual(sanitize(output));
  });

  it("should show unused components", async () => {
    const path = `test/yaml-filter-unused-components`
    const inputFile = `${path}/input.yaml`
    const outputFile = `${path}/output.yaml`
    const output = (fs.readFileSync(outputFile, 'utf8'));
    const setting = `${path}/customFilter.yaml`

    let result = await testUtils.cli([inputFile, `--filterFile ${setting}`, `--no-sort`, `-v`], '.');
    // console.log('result', result)
    expect(result.code).toBe(0);
    expect(result.stdout).toContain("formatted successfully");
    expect(result.stdout).toMatchSnapshot();
    expect(sanitize(result.stderr)).toStrictEqual(sanitize(output));
  });

  it("should use the rename", async () => {
    const path = `test/yaml-rename`
    const inputFile = `${path}/input.yaml`
    const outputFile = `${path}/output.yaml`
    const output = (fs.readFileSync(outputFile, 'utf8'));

    let result = await testUtils.cli([inputFile, `--rename "OpenAPI Petstore - OpenAPI 3.0"`, `--no-sort`], '.');
    // console.log('result', result)
    expect(result.code).toBe(0);
    expect(result.stdout).toContain("formatted successfully");
    expect(result.stdout).toMatchSnapshot();
    expect(sanitize(result.stderr)).toStrictEqual(sanitize(output));
  });

  it("should not convert large numbers in YAML", async () => {
    const path = `test/yaml-default-bug-big-numbers`
    const inputFile = `${path}/input.yaml`
    const outputFile = `${path}/output.yaml`
    const output = (fs.readFileSync(outputFile, 'utf8'));

    let result = await testUtils.cli([inputFile, `--no-sort`], '.');
    // console.log('result', result)
    expect(result.code).toBe(0);
    expect(result.stdout).toContain("formatted successfully");
    expect(result.stdout).toMatchSnapshot();
    expect(sanitize(result.stderr)).toStrictEqual(sanitize(output));
  });

  it("should not convert large numbers in JSON", async () => {
    const path = `test/json-default-bug-big-numbers`
    const inputFile = `${path}/input.json`
    const outputFile = `${path}/output.json`
    const output = (fs.readFileSync(outputFile, 'utf8'));

    let result = await testUtils.cli([inputFile, `--json`, `--no-sort`], '.');
    // console.log('result', result)
    expect(result.code).toBe(0);
    expect(result.stdout).toContain("formatted successfully");
    expect(result.stdout).toMatchSnapshot();
    expect(sanitize(result.stderr)).toStrictEqual(sanitize(output));
  });

})

/**
 * Sanitize the output of the CLI
 * @param {string} str
 * @returns {*|string}
 */
function sanitize(str) {
  return str.replace(/^\s+|\s+$/g, '').trim();
}
