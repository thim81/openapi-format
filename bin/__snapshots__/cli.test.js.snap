// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`openapi-format CLI command should not convert large numbers in JSON 1`] = `
"================================================================================
OpenAPI-Format CLI settings:
- Input file:		test/json-default-bug-big-numbers/input.json
[32m================================================================================[0m
[32mâœ…  OpenAPI formatted successfully[0m
[32m================================================================================[0m
"
`;

exports[`openapi-format CLI command should not convert large numbers in YAML 1`] = `
"================================================================================
OpenAPI-Format CLI settings:
- Input file:		test/yaml-default-bug-big-numbers/input.yaml
[32m================================================================================[0m
[32mâœ…  OpenAPI formatted successfully[0m
[32m================================================================================[0m
"
`;

exports[`openapi-format CLI command should output the help 1`] = `
"Usage: cli <file> [options]

Format an OpenAPI document by ordering, formatting and filtering fields.

Options:
  -o, --output <output>                      save the formatted OpenAPI file as JSON/YAML
  -s, --sortFile <sortFile>                  the file to specify custom OpenAPI fields ordering
  -c, --casingFile <casingFile>              the file to specify casing rules
  -f, --filterFile <filterFile>              the file to specify filter rules
  -c, --configFile <configFile>              the file with the OpenAPI-format CLI options
  --no-sort                                  don't sort the OpenAPI file
  --sortComponentsFile <sortComponentsFile>  the file with components to sort alphabetically
  --lineWidth <lineWidth>                    max line width of YAML output (default: -1)
  --rename <oaTitle>                         overwrite the title in the OpenAPI document
  --convertTo <oaVersion>                    convert the OpenAPI document to OpenAPI version 3.1
  --json                                     print the file to stdout as JSON
  --yaml                                     print the file to stdout as YAML
  --version                                  output the version number
  -v, --verbose                              verbosity that can be increased (default: 0)
  -h, --help                                 display help for command
"
`;

exports[`openapi-format CLI command should show unused components 1`] = `
"================================================================================
OpenAPI-Format CLI settings:
- Filter file:		test/yaml-filter-unused-components/customFilter.yaml
- Input file:		test/yaml-filter-unused-components/input.yaml
================================================================================
Removed unused components:
- components/schemas		 \\"Customer\\"
- components/schemas		 \\"Unused\\"
- components/schemas		 \\"ErrorModel\\"
- components/schemas		 \\"Address\\"
- components/responses		 \\"GeneralError\\"
- components/parameters		 \\"offsetParam\\"
- components/parameters		 \\"limitParam\\"
- components/examples		 \\"tshirt-unused\\"
- components/examples		 \\"tshirt\\"
- components/requestBodies	 \\"NewItem\\"
- components/requestBodies	 \\"Pet\\"
- components/requestBodies	 \\"UserArray\\"
- components/requestBodies	 \\"UnusedArray\\"
- components/headers		 \\"X-RateLimit-Unused\\"
- components/headers		 \\"X-RateLimit-Limit\\"
- components/headers		 \\"X-RateLimit-Remaining\\"
Total components removed: 16
[32m================================================================================[0m
[32mâœ…  OpenAPI filtered & formatted successfully[0m
[32m================================================================================[0m
"
`;

exports[`openapi-format CLI command should stop and show error 1`] = `
"================================================================================
OpenAPI-Format CLI settings:
- Sort file:		foobar
[31m Sort file error - no such file or directory \\"foobar\\"
"
`;

exports[`openapi-format CLI command should use the casingFile 1`] = `
"================================================================================
OpenAPI-Format CLI settings:
- Casing file:		test/yaml-casing/customCasing.yaml
- Input file:		test/yaml-casing/input.yaml
[32m================================================================================[0m
[32mâœ…  OpenAPI formatted successfully[0m
[32m================================================================================[0m
"
`;

exports[`openapi-format CLI command should use the convert version 1`] = `
"================================================================================
OpenAPI-Format CLI settings:
- Input file:		test/yaml-convert-3.1/input.yaml
- OAS version converted to: \\"3.1\\"
================================================================================

OpenAPI-Format CLI options:
|-----------|-------|
| sort      | false |
| lineWidth | -1    |
| convertTo | 3.1   |
| verbose   | 3     |
|-----------|-------|

[32m================================================================================[0m
[32mâœ…  OpenAPI formatted successfully[0m
[32m================================================================================[0m
"
`;

exports[`openapi-format CLI command should use the default sort 1`] = `
"================================================================================
OpenAPI-Format CLI settings:
- Sort file:		(defaultSort.json)
- Input file:		test/yaml-default/input.yaml
[32m================================================================================[0m
[32mâœ…  OpenAPI formatted successfully[0m
[32m================================================================================[0m
"
`;

exports[`openapi-format CLI command should use the filterFile 1`] = `
"================================================================================
OpenAPI-Format CLI settings:
- Filter file:		test/yaml-filter-custom/customFilter.yaml
- Input file:		test/yaml-filter-custom/input.yaml
[32m================================================================================[0m
[32mâœ…  OpenAPI filtered & formatted successfully[0m
[32m================================================================================[0m
"
`;

exports[`openapi-format CLI command should use the no-sort 1`] = `
"================================================================================
OpenAPI-Format CLI settings:
- Input file:		test/yaml-no-sort/input.yaml
[32m================================================================================[0m
[32mâœ…  OpenAPI formatted successfully[0m
[32m================================================================================[0m
"
`;

exports[`openapi-format CLI command should use the rename 1`] = `
"================================================================================
OpenAPI-Format CLI settings:
- Input file:		test/yaml-rename/input.yaml
[32m================================================================================[0m
[32mâœ…  OpenAPI formatted successfully[0m
[32m================================================================================[0m
"
`;

exports[`openapi-format CLI command should use the sortComponentsFile 1`] = `
"================================================================================
OpenAPI-Format CLI settings:
- Sort file:		(defaultSort.json)
- Sort Components file:	test/yaml-sort-components/customSortComponents.yaml
- Input file:		test/yaml-sort-components/input.yaml
[32m================================================================================[0m
[32mâœ…  OpenAPI formatted successfully[0m
[32m================================================================================[0m
"
`;

exports[`openapi-format CLI command should use the sortFile 1`] = `
"================================================================================
OpenAPI-Format CLI settings:
- Sort file:		test/yaml-custom/customSort.yaml
- Input file:		test/yaml-custom/input.yaml
[32m================================================================================[0m
[32mâœ…  OpenAPI formatted successfully[0m
[32m================================================================================[0m
"
`;
